pkg/operator/ceph/controllers/clusterdisruption/osd.go

```diff
@@ -0,0 +1,207 @@
+	mux                sync.Mutex
+}
+
+type cachedOSDLocation struct {
+	result     *cephClient.CrushFindResult
+	lastSynced time.Time
+}
+
+// totest
+//Get takes an osd id and returns a CrushFindResult from cache
+func (o *OSDCrushLocationMap) Get(clusterNamespace string, id int) (*cephClient.CrushFindResult, error) {
+	o.mux.Lock()
+	defer o.mux.Unlock()
+	if o.ResyncPeriod == 0 {
+		o.ResyncPeriod = 30 * time.Minute
+	}
+
+	// initialize clusterLocationMap
+	if len(o.clusterLocationMap) == 0 {
+		o.clusterLocationMap = make(map[string]map[int]cachedOSDLocation)
+	}
+	locationMap, ok := o.clusterLocationMap[clusterNamespace]
+	// initialize namespace map
+	if !ok {
+		o.clusterLocationMap[clusterNamespace] = make(map[int]cachedOSDLocation)
+	}
+
+	// sync of osd id not found in clusterNamespace
+	osdLocation, ok := locationMap[id]
+	if !ok {
+		osdResult, err := o.get(clusterNamespace, id)
+		if err != nil {
+			return nil, fmt.Errorf("failed to run `find` on osd %d in cluster %s", id, clusterNamespace)
+		}
+		o.clusterLocationMap[clusterNamespace][id] = cachedOSDLocation{result: osdResult, lastSynced: time.Now()}
+		return osdResult, nil
+	}
+
+	// sync if not synced for longer than ResyncPeriod
+	if time.Since(osdLocation.lastSynced) > o.ResyncPeriod {
+		osdResult, err := o.get(clusterNamespace, id)
+		if err != nil {
+			return nil, fmt.Errorf("failed to run `find` on osd %d in cluster %s", id, clusterNamespace)
+		}
+		o.clusterLocationMap[clusterNamespace][id] = cachedOSDLocation{result: osdResult, lastSynced: time.Now()}
+		return osdResult, nil
+	}
+
+	return osdLocation.result, nil
+
+}
+
+// uncached version
+func (o *OSDCrushLocationMap) get(clusterNamespace string, id int) (*cephClient.CrushFindResult, error) {
+	osdResult, err := cephClient.FindOSDInCrushMap(o.Context, clusterNamespace, id)
+	if err != nil {
+		return nil, fmt.Errorf("failed running find on osd %d: %+v", id, err)
+	}
+	o.clusterLocationMap[clusterNamespace][id] = cachedOSDLocation{
+		result:     osdResult,
+		lastSynced: time.Now(),
+	}
+	return osdResult, nil
+}
+
+func getOSDsForNodes(osdDataList []OsdData, nodeList []*corev1.Node) ([]OsdData, error) {
+	nodeOsdDataList := make([]OsdData, 0)
+	for _, node := range nodeList {
+		if node == nil {
+			logger.Warningf("node in nodelist was nil")
+			continue
+		}
+		nodeLabels := node.ObjectMeta.GetLabels()
+		for _, osdData := range osdDataList {
+			crushHostname, ok := osdData.CrushMeta.Location["host"]
+			if !ok {
+				return nil, fmt.Errorf("could not find the CrushFindResult.Location['host'] for %s", osdData.Deployment.ObjectMeta.Name)
+			}
+			nodeHostname, ok := nodeLabels[corev1.LabelHostname]
+			if !ok {
+				return nil, fmt.Errorf("could not find the %s label on node %s", corev1.LabelHostname, node.ObjectMeta.Name)
+			}
+			// TODO: implement checking for virtual nodes if the osds are portable.
+			// virtual node implementation for portable osds is pending. https://github.com/rook/rook/issues/3648
```

Comment:
When we discussed this with sage, he had mentioned that even if PVCs are portable, we need to put them in the host type because small things in ceph break if "host" is not populated.
I assumed that the PVC name was being put in the crush host name category so this would tell us which zone/region each OSD was under. I'll just double-check that assumption.

IMO, we will need a virtual node if we have sub-zone failure domains with portable OSDs or data will not be spread across failure domains.

Reply:
Looks like it won't work for OSDs on PVs, small fix needed, still depending on crush map. Can't support sub-zone failure domain in that scenario.

Reply:
Updating to check if they share a failure domain OSD location is derived from CRUSH and Node location is derived from Topology Labels.

Still will not work on OSDs on PVs until topology aware is fixed for them.